### PR TITLE
test: harden pulse prefetch regression harness

### LINKED ISSUE #26764: quality-debt: .agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh — PR #26762 review feedback (medium)
## Unactioned Review Feedback

**Source PR**: #26762
**File**: `.agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh`
**Reviewers**: gemini
**Findings**: 5
**Max severity**: medium

---

### MEDIUM: gemini (gemini-code-assist[bot])
**File**: `.agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh:13`
medium

Under `set -u`, referencing `${BASH_SOURCE[0]}` directly can cause an unbound variable error and terminate the script if it is executed in an environment where `BASH_SOURCE` is not set (for example, when the script is piped directly into `bash` via stdin).

Using the parameter expansion `${BASH_SOURCE[0]:-$0}` safely falls back to `$0` (which is always set) and prevents premature script termination.

```suggestion
TEST_SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]:-$0}")" && pwd)"
```

<details>
<summary>References</summary>

1. In Shell scripts with `set -u` or `set -euo pipefail` enabled, referencing a potentially unset variable directly (such as in `[[ ! -f "$VAR" ]]`) will cause an unbound variable error and terminate the script. Use parameter expansion `${VAR:-}` to safely check if the variable is empty or unset before performing file or other checks.
</details>

[View comment](https://github.com/marcusquinn/aidevops/pull/26762#discussion_r3533398445)

---

### MEDIUM: gemini (gemini-code-assist[bot])
**File**: `.agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh:17`
medium

Under `set -u`, referencing `$1` directly when the function is called without arguments will trigger an unbound variable error and terminate the script.

Using parameter expansion `${1:-}` safely defaults to an empty string if no argument is provided.

```suggestion
	local message="${1:-}"
```

<details>
<summary>References</summary>

1. When referencing optional positional parameters in Bash scripts where `set -u` (nounset) is active, use safe parameter expansion fallbacks (e.g., `${3:-${2:-}}`) to prevent unbound variable errors when the parameters are omitted by the caller.
</details>

[View comment](https://github.com/marcusquinn/aidevops/pull/26762#discussion_r3533398450)

---

### MEDIUM: gemini (gemini-code-assist[bot])
**File**: `.agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh:23`
medium

Under `set -u`, referencing `$1` directly when the function is called without arguments will trigger an unbound variable error and terminate the script.

Using parameter expansion `${1:-}` safely defaults to an empty string if no argument is provided.

```suggestion
	local message="${1:-}"
```

<details>
<summary>References</summary>

1. When referencing optional positional parameters in Bash scripts where `set -u` (nounset) is active, use safe parameter expansion fallbacks (e.g., `${3:-${2:-}}`) to prevent unbound variable errors when the parameters are omitted by the caller.
</details>

[View comment](https://github.com/marcusquinn/aidevops/pull/26762#discussion_r3533398452)

---

### MEDIUM: gemini (gemini-code-assist[bot])
**File**: `.agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh:34`
medium

Checking for executable permissions (`-x`) can sometimes fail in environments where file permissions are not fully preserved (e.g., certain CI/CD runners, Windows checkouts, or archive extractions).

Since `pulse-unbound-var-check.sh` is a shell script, we can check for file existence (`-f`) instead and execute it by passing it directly to `bash`. This is more portable and robust.

```suggestion
	if [[ ! -f "$checker" ]]; then
```

<details>
<summary>References</summary>

1. When checking for the presence of a script that is executed by passing it as an argument to an interpreter (such as `bun` or `node`), use the `-f` flag instead of `-x` in shell scripts, as the file does not require the executable bit to be set.
</details>

[View comment](https://github.com/marcusquinn/aidevops/pull/26762#discussion_r3533398459)

---

### MEDIUM: gemini (gemini-code-assist[bot])
**File**: `.agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh:39`
medium

Execute the script by passing it directly to `bash` rather than executing it directly, ensuring compatibility even if the executable bit is not set.

```suggestion
	output=$(bash "$checker" --scan-files "$orchestration" "$repo" 2>&1) || {
```

<details>
<summary>References</summary>

1. When checking for the presence of a script that is executed by passing it as an argument to an interpreter (such as `bun` or `node`), use the `-f` flag instead of `-x` in shell scripts, as the file does not require the executable bit to be set.
</details>

[View comment](https://github.com/marcusquinn/aidevops/pull/26762#discussion_r3533398461)

---

---

_Auto-generated by `quality-feedback-helper.sh scan-merged`. Review each finding and either fix the code or dismiss with a reason._

---
[aidevops.sh](https://aidevops.sh) v3.31.80 automated scan.

### PR DESCRIPTION
**Pull request by alex-solovyev**
## Summary
- Harden `.agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh` against nounset edge cases in the script path and helper messages.
- Check the checker script by file existence and invoke it through `bash` so executable permissions are not required.

Resolves #26764

## Testing
- `bash .agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh`
- `shellcheck .agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh`

---
[aidevops.sh](https://aidevops.sh) v3.31.80 plugin for [OpenCode](https://opencode.ai) v1.17.14 with gpt-5.5 spent 1m and 34,465 tokens on this as a headless worker.

### CODE REVIEW

**Comment by alex-solovyev** [comment]
## Merge summary

- Hardened `.agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh` against `set -u` edge cases by using safe fallbacks for `BASH_SOURCE[0]` and helper message parameters.
- Made the regression harness portable across checkouts without executable-bit preservation by checking `-f` and invoking the checker through `bash`.

## Verification

- `bash .agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh`
- `shellcheck .agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh`

---
[aidevops.sh](https://aidevops.sh) v3.31.80 plugin for [OpenCode](https://opencode.ai) v1.17.14 with gpt-5.5 spent 2m and 35,536 tokens on this as a headless worker.

**Review by marcusquinn — APPROVED** [review]
Auto-approved by pulse runner @marcusquinn — author @alex-solovyev confirmed collaborator, pre-merge gates passed.

---
[aidevops.sh](https://aidevops.sh) v3.31.80 automated scan.

File: .agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh [hunk 1]
```diff
@@ -10,17 +10,17 @@
 
 set -u
 
-TEST_SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+TEST_SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]:-$0}")" && pwd)"
```
[the remark this answers is not in the document]
Comment (marcusquinn) [reply .agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh]: @gemini-code-assist 
Verified on the current PR head: this outdated thread is already addressed. `.agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh:13` now uses `${BASH_SOURCE:-$0}`, and the focused regression test plus ShellCheck pass:

- `.agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh`
- `shellcheck .agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh`

File: .agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh [hunk 2]
```diff
@@ -31,12 +31,12 @@ main() {
...
 
-	if [[ ! -x "$checker" ]]; then
-		fail "pulse unbound-var checker is not executable: ${checker}"
+	if [[ ! -f "$checker" ]]; then
+		fail "pulse unbound-var checker is missing: ${checker}"
 		return 1
 	fi
 
-	output=$("$checker" --scan-files "$orchestration" "$repo" 2>&1) || {
+	output=$(bash "$checker" --scan-files "$orchestration" "$repo" 2>&1) || {
```
[the remark this answers is not in the document]
Comment (marcusquinn) [reply .agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh]: @gemini-code-assist 
Verified on the current PR head: this outdated thread is already addressed. `.agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh:29` defines `bash_bin="${BASH:-bash}"`, and line 40 invokes the checker through that interpreter. Focused regression test and ShellCheck pass:

- `.agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh`
- `shellcheck .agents/scripts/tests/test-pulse-prefetch-unbound-var-regression.sh`

**Comment by marcusquinn** [comment]
REVIEW_FOLLOWUP_CREATED source_pr=26780 issue=26791 fingerprint=source-pr-26780 runner=marcusquinn ts=2026-07-11T15:05:54Z


---
[aidevops.sh](https://aidevops.sh) v3.32.21 automated scan.
